src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/BoundPropertyManager.cs

```diff
@@ -0,0 +1,246 @@
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            allowedSources = allowedSources ?? DefaultAllowedSources;
+
+            var tempDataPrefix = options.SuppressTempDataPropertyPrefix ?
+                string.Empty :
+                "TempDataProperty-";
+
+            var properties = GetProperties(type, allowedSources, tempDataPrefix, requireViewDataPropertySetters);
+            return new BoundPropertyManager(properties ?? Array.Empty<PropertyItem>());
+        }
+
+        public void Populate(object instance, BoundPropertyContext context)
+        {
+            if (instance == null)
+            {
+                throw new ArgumentNullException(nameof(instance));
+            }
+
+            for (var i = 0; i < PropertyItems.Count; i++)
+            {
+                var propertyItem = PropertyItems[i];
+                var propertyName = propertyItem.PropertyHelper.Name;
+
+                object value;
+                switch (propertyItem.Source)
+                {
+                    case BoundPropertySource.TempData:
+                        value = context.TempData[propertyItem.SourceKey];
+                        break;
+                    case BoundPropertySource.ViewData:
+                        value = context.ViewData[propertyItem.SourceKey];
+                        break;
+                    default:
+                        throw new InvalidOperationException(Resources.FormatUnsupportedEnumType(propertyItem.Source));
+                }
+
+                if (value == null)
+                {
+                    continue;
+                }
+
+                propertyItem.PropertyHelper.SetValue(instance, value);
+            }
+        }
+
+        public void Save(object instance, BoundPropertyContext context)
+        {
+            if (instance == null)
+            {
+                throw new ArgumentNullException(nameof(instance));
+            }
+
+            for (var i = 0; i < PropertyItems.Count; i++)
+            {
+                var propertyItem = PropertyItems[i];
+                var key = propertyItem.SourceKey;
+
+                var currentValue = propertyItem.PropertyHelper.GetValue(instance);
+                if (propertyItem.Source == BoundPropertySource.TempData)
+                {
+                    var originalValue = context.TempData[key];
+                    if (currentValue != null && !currentValue.Equals(originalValue))
+                    {
+                        context.TempData[key] = currentValue;
+                        // Mark the key to be kept. This ensures that even if something later in the execution pipeline reads it,
+                        // such as another view with a `TempData` property, the key is preserved through the current request.
+                        context.TempData.Keep(key);
```

Comment:
We compare the value in TempDataDictionary to the current property value. If it's changed, we preserve it for the current request. Does this miss any important scenarios that we accounted for in the past?